File: rn-compiler/src/main/java/com/ldy/compiler/RNProcessor.java
package com.ldy.compiler;

import com.ldy.annotation.ReactMethod;
import com.ldy.annotation.ReactSyncHook;

import java.util.HashSet;
import java.util.Set;
import java.util.function.Consumer;

import javax.annotation.processing.AbstractProcessor;
import javax.annotation.processing.Filer;
import javax.annotation.processing.Messager;
import javax.annotation.processing.ProcessingEnvironment;
import javax.annotation.processing.RoundEnvironment;
import javax.lang.model.SourceVersion;
import javax.lang.model.element.Element;
import javax.lang.model.element.TypeElement;
import javax.lang.model.util.Elements;

/**
 * Created by endsmile on 2018/3/8.
 */

public class RNProcessor extends AbstractProcessor {
    private Elements elementUtils;
    private Filer filer;
    private Messager messager;

    @Override
    public void init(ProcessingEnvironment env) {
        super.init(env);
        elementUtils = env.getElementUtils();
        filer = env.getFiler();
        messager = env.getMessager();

    }
    
    @Override
    public boolean process(Set<? extends TypeElement> set, RoundEnvironment roundEnvironment) {

        return false;
    }

    @Override
    public Set<String> getSupportedAnnotationTypes(){
        return new HashSet<String>(){{
            add(ReactMethod.class.getCanonicalName());
            add(ReactSyncHook.class.getCanonicalName());
        }};
    }


    @Override
    public SourceVersion getSupportedSourceVersion(){
        return SourceVersion.latestSupported();
    }
}

File: compiler/src/main/java/com/ldy/compiler/RNProcessor.kt
package com.ldy.compiler

import com.grosner.kpoet.*
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReactSyncHook
import com.squareup.javapoet.ParameterizedTypeName
import java.util.*
import javax.annotation.processing.*
import javax.lang.model.SourceVersion
import javax.lang.model.element.TypeElement
import javax.lang.model.util.Elements
import kotlin.collections.HashSet

/**
 * Created by ldy on 2017/5/11.
 */

class RNProcessor : AbstractProcessor() {

    private lateinit var elementUtils: Elements
    private lateinit var filer: Filer
    private lateinit var messager: Messager

    @Synchronized override fun init(env: ProcessingEnvironment) {
        super.init(env)
        elementUtils = env.elementUtils
        filer = env.filer
        messager = env.messager

    }

    override fun process(set: Set<TypeElement>, roundEnvironment: RoundEnvironment): Boolean {
        roundEnvironment.getElementsAnnotatedWith(ReactMethod::class.java)
                .plus(roundEnvironment.getElementsAnnotatedWith(ReactSyncHook::class.java))
                .groupBy { it.enclosingElement as TypeElement }
                .forEach { classElement, annotationList ->
                    val packageName = elementUtils.getPackageOf(classElement).qualifiedName.toString()
                    val reactMethodList = annotationList.filter { it.getAnnotation(ReactMethod::class.java) != null }
                    val reactSyncHookList = annotationList.filter { it.getAnnotation(ReactSyncHook::class.java) != null }
                    javaFile(packageName){
                        `class`(classElement.simpleName.toString()+"\$\$Assist") {
                            modifiers(public, final)

                            `public static final field`(ParameterizedTypeName.get(Set::class.java,String::class.java), "reactMethodSet"){
                                javadoc("Don't modify it,otherwise will throw exception")
                            }
                            `public static final field`(ParameterizedTypeName.get(Set::class.java,String::class.java), "reactSyncHookSet"){
                                javadoc("Don't modify it,otherwise will throw exception")
                            }

                            `static block`{
                                statement("\$T<\$T> set1 = new \$T<\$T>(${reactMethodList.size});",Set::class.java,String::class.java,HashSet::class.java,String::class.java)
                                reactMethodList
                                        .map { it.simpleName.toString() }
                                        .forEach { statement("set1.add(${it.S})") }
                                statement("reactMethodSet = \$T.unmodifiableSet(set1)",Collections::class.java)

                                add("\n")

                                statement("\$T<\$T> set2 = new \$T<\$T>(${reactSyncHookList.size});",Set::class.java,String::class.java,HashSet::class.java,String::class.java)
                                reactSyncHookList
                                        .map { it.simpleName.toString() }
                                        .forEach { statement("set2.add(${it.S})") }
                                statement("reactSyncHookSet = \$T.unmodifiableSet(set2)",Collections::class.java)
                            }

                        }
                    }.writeTo(filer)
                }

        return true
    }


    override fun getSupportedAnnotationTypes(): Set<String>
            = hashSetOf(ReactMethod::class.java.canonicalName, ReactSyncHook::class.java.canonicalName)


    override fun getSupportedSourceVersion(): SourceVersion {
        return SourceVersion.latestSupported()
    }

}
